src/traverse.py

```diff
@@ -12,38 +15,32 @@
 env = os.environ.get("INPUT_ENVIRONMENT")
 root_dir = os.environ.get("INPUT_DIRECTORY")
 pattern = os.environ.get("INPUT_PATTERN")
+chart_dir = os.environ.get("INPUT_CHART")
 rendered_path_input = os.environ.get("INPUT_RENDERED_PATH", "rendered").strip()
-print("Rendered path: " + rendered_path_input)
-output = []
-if env and root_dir and pattern:
-    pattern_reg = re.compile(pattern)  # Match all files ending with input pattern"
-    logging.info("Pattern regex: %s", pattern_reg)
-    for item in os.walk(root_dir):
-        # print(item[0])
-        item_path = Path(item[0])
-        env_item_path = item_path.absolute().as_posix()+ "/" + env
-        rendered_path = item_path.parent.absolute().as_posix() + "/" + rendered_path_input
-        
-        # logging.info("Checking file: " + item[0]+ " " + item[1]) 
-        if pattern_reg.search(env_item_path) and os.path.exists(rendered_path):
-            
-            logging.info("Found dir: " + env_item_path + " " + rendered_path)
-            jobs_list = os.listdir(env_item_path)
-            logging.info("Jobs list: " + str(jobs_list))
-            for job in jobs_list:
-                base_job_path = item_path.absolute().as_posix() + "/base/" + job
-                job_path = env_item_path + "/" + job
-                logging.info("Job path: " + job_path)
-                job_rendered_path = rendered_path + "/"+env  +"/"  + job
-                logging.info("Job rendered path: " + job_rendered_path)
-                if os.path.exists(job_rendered_path):
-                    logging.info("Job rendered path exists: " + job_rendered_path)
-                    output.append(job)
-                else:
-                    logging.info("Job rendered path does not exist: " + job_rendered_path)
-            output.append(item[0])
-else:
-    exit('ERROR: the INPUT_NUM or INPUT_ENVIRONMENT is not provided')
+# Check if required inputs are provided
+if not env or not root_dir or not pattern or not chart_dir:
+    logging.error("Missing required input")
+    sys.exit(1)
+PATTERN_REG = re.compile(pattern)  # Match all files ending with input pattern"
 
+output = list()
+for root, _dirs, _files in os.walk(root_dir):
+    item_path = Path(root)
+    env_item_path = item_path.absolute().as_posix()
+    jobs_list = os.listdir(env_item_path)
+    rendered_path = f"{item_path.parent.absolute().as_posix()
+                       }/{rendered_path_input}"
+    if PATTERN_REG.search(env_item_path) and os.path.exists(rendered_path):
+        logging.info("Found dir: %s %s", env_item_path, rendered_path)
+        for job in jobs_list:
```

Comment:
instead of doing another for loop we can instead put all the results of your found path and put them into an array, then create another for loop outside of current for loop and loop array, this way you're doing o(n) instead of o(n^2)
```
found_paths = []
for root, _, _, in os.walk(root_dir):
...
  found_paths.append(path)
for path in found_paths:
  ...do stuff
```